test/model/internal_link_test.dart

```diff
@@ -0,0 +1,388 @@
+  final store = eg.store(account: account);
+  if (streams != null) {
+    store.addStreams(streams);
+  }
+  store.addUser(eg.selfUser);
+  if (users != null) {
+    store.addUsers(users);
+  }
+  return store;
+}
+
+void main() {
+  final realmUrl = Uri.parse('https://example.com/');
+
+  testExpectedNarrows(List<(String, Narrow?)> testCases, {
+    List<ZulipStream>? streams,
+    PerAccountStore? store,
+    List<User>? users,
+  }) {
+    assert((streams != null || users != null) ^ (store != null));
+    store ??= setupStore(realmUrl: realmUrl, streams: streams, users: users);
+    for (final testCase in testCases) {
+      final String urlString = testCase.$1;
+      final Uri url = tryResolveOnRealmUrl(urlString, realmUrl)!;
+      final Narrow? expected = testCase.$2;
+      test(urlString, () {
+        check(parseInternalLink(url, store!)).equals(expected);
+      });
+    }
+  }
+
+  group('parseInternalLink', () {
+    final streams = [
+      eg.stream(streamId: 1, name: 'check'),
+    ];
+    final testCases = [
+      (true, 'legacy: stream name, no ID',
+        '#narrow/stream/check', realmUrl),
+      (true, 'legacy: stream name, no ID, topic',
+        '#narrow/stream/check/topic/topic1', realmUrl),
+
+      (true, 'with numeric stream ID',
+        '#narrow/stream/123-check', realmUrl),
+      (true, 'with numeric stream ID and topic',
+        '#narrow/stream/123-a/topic/topic1', realmUrl),
+
+      (true, 'with numeric pm user IDs (new operator)',
+        '#narrow/dm/123-mark', realmUrl),
+      (true, 'with numeric pm user IDs (old operator)',
+        '#narrow/pm-with/123-mark', realmUrl),
+
+      (false, 'wrong fragment',
+        '#nope', realmUrl),
+      (false, 'wrong path',
+        'user_uploads/#narrow/stream/check', realmUrl),
+      (false, 'wrong domain',
+        'https://another.com/#narrow/stream/check', realmUrl),
+
+      (false, '#narrowly',
+        '#narrowly/stream/check', realmUrl),
+
+      (false, 'double slash',
+        'https://example.com//#narrow/stream/check', realmUrl),
+      (false, 'triple slash',
+        'https://example.com///#narrow/stream/check', realmUrl),
+
+      (true, 'with port',
+        'https://example.com:444/#narrow/stream/check',
+        Uri.parse('https://example.com:444/')),
+
+      // Dart's [Uri] currently lacks IDNA or Punycode support:
+      //   https://github.com/dart-lang/sdk/issues/26284
+      //   https://github.com/dart-lang/sdk/issues/29420
+
+      // (true, 'same domain, punycoded host',
+      //   'https://example.xn--h2brj9c/#narrow/stream/check'),
+      //   Uri.parse('https://example.भारत/')), // FAILS
+
+      (true, 'punycodable host',
+        'https://example.भारत/#narrow/stream/check',
+        Uri.parse('https://example.भारत/')),
+
+      // (true, 'same domain, IDNA-mappable'
+      //   'https://ℯⅩªm🄿ₗℰ.ℭᴼⓂ/#narrow/stream/check'),
+      //   Uri.parse('https://example.com'), // FAILS
```

Comment:
This version doesn't parse if you uncomment it.

Looking at the neighboring commented-out failures, they don't either (but more subtly) — there's a stray close-paren.